README.md
# timeit2 


このパッケージは処理時間の計測module [timeit](https://docs.python.org/3/library/timeit.html) を使いやすくしたものです。以下の特徴があります。

- 引数を使い回しません(timeitは引数を書き換える処理をした時にそのまま次に渡します)
- 複数の関数を1度に計測して相対的な速度差を視覚化できます
- 処理時間の桁数を変えられます(0.00374567より0.0037の方が見やすい)
- 返り値を一緒にprintできます

![test](https://github.com/atu4403/timeit2/workflows/test/badge.svg)


## Install

```
$ pip install timeit2
```


## Example

```python
from timeit2 import ti2

arg = range(10 ** 6)
ti2(
    max,
    sum,
    args=[arg],
    floor=4,
    print_return=True,
    return_label="ret: ",
    relative=True,
)
# max:
#         0.0356 sec
#         ret: 999999
# sum:
#         0.0249 sec
#         ret: 499999500000
# relative:
#         sum:
#                 1
#         max:
#                 1.43
```

最小限の使い方

```python
from timeit2 import ti2

def f():
    li = []
    for i in range(10 ** 5):
        li.append(i * i)
    return li

ti2(f)

# f:
#         0.015094 sec
```

配列作成に`list.append`,`list.insert`,`deque`をそれぞれ使った処理時間と相対評価

```python
from collections import deque
from timeit2 import ti2


def append_(n):
    li = []
    for i in range(n):
        li.append(i)
    return li


def insert_(n):
    li = []
    for i in range(n):
        li.insert(0, i)
    return li


def deque_right(n):
    li = deque()
    for i in range(n):
        li.append(i)
    return li


def deque_left(n):
    li = deque()
    for i in range(n):
        li.appendleft(i)
    return li


ti2(
    append_,
    insert_,
    deque_right,
    deque_left,
    args=[10 ** 4],
    relative=True,
)

# append_:
#         0.001118 sec
# insert_:
#         0.021187 sec
# deque_right:
#         0.000891 sec
# deque_left:
#         0.000855 sec
# relative:
#         deque_left:
#                 1
#         deque_right:
#                 1.04
#         append_:
#                 1.31
#         insert_:
#                 24.77
```

## API

```python
ti2(
    *fn: Callable,
    args: list = [],
    number: int = 100,
    floor: int = 6,
    relative: bool = False,
    print_return: bool = False,
    return_label: str = "",
)
```

### argument

#### fn

Type: Callable

計測する関数、可変長引数なので複数指定が可能。
引数がある場合は`args=` で指定する。

### Options

#### args 

Type: list

Default: []

fnに渡す引数のlist

#### number 

Type: int

Default: 100

各関数を計測する回数、結果はその最速値が表示される


#### floor

Type: int

Default: 6

表示する小数点以下の桁数


#### relative

Type: bool

Default: False

計測結果に相対評価を追加する

#### print_return

Type: bool

Default: False

計測結果に関数の返り値を追加する

#### return_label

Type: str
Default: ""

`print_return=True`の場合、返り値のラベルを指定する


## alias

ti2, ti, timeit, timeit2 をimportできますが、全て同じものです。


```python
from timeit2 import ti2, ti, timeit, timeit2
```

## Related
- [timeit](https://docs.python.org/3/library/timeit.html)


## License

MIT © 2021 [atu4403](https://github.com/atu4403)


CHANGELOG.md

<a name="v1.0.0"></a>
## [v1.0.0](/compare/v0.1.1...v1.0.0)

> 2021-12-28

major release


tests/test_timeit2.py
from pytest import approx
import types
from timeit2 import ti2, ti, timeit, timeit2


def fn1(n):
    for i in range(10 ** 4):
        pass
    return 1


def fn2(n):
    return 1


def fn3(n):
    for i in range(10 ** 3):
        pass
    return 1.2


def test_full_args(mocker):
    mocker.patch("builtins.print")
    f1 = mocker.MagicMock(side_effect=fn1)
    f2 = mocker.MagicMock(side_effect=fn2)
    f3 = mocker.MagicMock(side_effect=fn3)
    f1.__name__ = "f1"
    f2.__name__ = "f2"
    f3.__name__ = "f3"
    a = ti2(
        f1,
        f2,
        f3,
        args=[9],
        number=1000,
        print_return=True,
        floor=7,
        return_label="return: ",
        relative=True,
    )
    assert f1.call_count == 1000
    assert f2.call_count == 1000
    assert f3.call_count == 1000
    assert print.call_count == 9
    assert a[0][1] <= a[1][1] <= a[2][1]
    assert a[0][0] == "f2" and a[1][0] == "f3" and a[2][0] == "f1"


def test_number(mocker):
    mocker.patch("builtins.print")
    ti(fn1, args=[100], number=1)
    assert print.call_count == 1


def test_alias(mocker):
    assert ti == ti2 == timeit == timeit2


def test_return(mocker):
    mocker.patch("builtins.print")
    result = ti2(fn1, fn2, fn3, args=[1])
    assert type(result) == list
    for tup in result:
        assert tup[0] in ["fn1", "fn2", "fn3"]
        assert type(tup[1]) == float


src/timeit2/__init__.py
import time
import copy
from typing import Callable


def ti2(
    *fn: Callable,
    args: list = [],
    number: int = 100,
    floor: int = 6,
    relative: bool = False,
    print_return: bool = False,
    return_label: str = "",
):
    """Extended version of timeit package

    Args:
        fn (Callable): Time measuring function
        args (list, optional): argument of fn. Defaults to [].
        number (int, optional): Time measuring count number. Defaults to 100.
        floor (int, optional): Number after the decimal point. Defaults to 6.
        relative (bool, optional): Print relative evaluation. Defaults to False.
        print_return (bool, optional): Print the return value. Defaults to False.
        return_label (str, optional): Return value label. Defaults to "".

    Returns:
        list[tuple]: list of measured times
    """
    times = []
    for f in fn:
        fres = {"name": f.__name__, "results": [], "times": []}
        for _ in range(number):
            _args = copy.deepcopy(args)
            start = time.perf_counter()
            fres["results"].append(f(*_args))
            fres["times"].append(time.perf_counter() - start)
        t = fres["times"]
        fres["fast"] = min(t)
        times.append(
            (
                fres["name"],
                fres["fast"],
            )
        )
        print(f"{fres['name']}:\n\t{fres['fast']:.{floor}f} sec")
        if print_return:
            r = fres["results"][0]
            print(f"\t{return_label}{fres['results'][0]}")
    if relative:
        times.sort(key=lambda x: x[1])
        first = True
        base = -1.0
        for t in times:
            if first:
                base = t[1]
                print(f"relative:\n\t{t[0]}:\n\t\t1")
                first = False
            else:
                r = t[1] / base
                print(f"\t{t[0]}:\n\t\t{r:.2f}")
    return times


ti = ti2
timeit = ti2
timeit2 = ti2


examples/append.py
from collections import deque
from timeit2 import ti2


def append_(n):
    li = []
    for i in range(n):
        li.append(i)
    return li


def insert_(n):
    li = []
    for i in range(n):
        li.insert(0, i)
    return li


def deque_right(n):
    li = deque()
    for i in range(n):
        li.append(i)
    return li


def deque_left(n):
    li = deque()
    for i in range(n):
        li.appendleft(i)
    return li


ti2(
    append_,
    insert_,
    deque_right,
    deque_left,
    args=[10 ** 4],
    relative=True,
)

# append_:
#         0.001118 sec
# insert_:
#         0.021187 sec
# deque_right:
#         0.000891 sec
# deque_left:
#         0.000855 sec
# relative:
#         deque_left:
#                 1
#         deque_right:
#                 1.04
#         append_:
#                 1.31
#         insert_:
#                 24.77
